### ISSUE TITLE
Set patterns

### ISSUE THREAD

**Issue by miDeb** [dart-lang/language#3296]
Currently, there are Map patterns and List patterns, but no Set patterns. I haven't found an issue tracking this, and I believe this would be a nice feature.

For example, in flutter there is `MaterialStateProperty.resolveWith`, which givies you a set of states that you have to resolve to a certain value (such as a color).
Example usage:
```dart
MaterialStateProperty.resolveWith((states) {
  return switch(states) {
    {MaterialState.selected, MaterialState.disabled} => Colors.grey,
    {MaterialState.selected} => Colors.yellow,
    _ => Colors.transparent,
  };
}
```

**Comment by mateusfccp** [comment]
Currently, you can do this workaround:
```dart
MaterialStateProperty.resolveWith((states) {
  return switch([...states]) {
    [MaterialState.selected, MaterialState.disabled] => Colors.grey,
    [MaterialState.selected] => Colors.yellow,
    _ => Colors.transparent,
  };
}
```

There's a small overhead from creating a new list, but shouldn't be a problem.

**Comment by miDeb** [comment]
I don't think this would work reliably, as there is no guarantee that `MaterialState.selected` will come before `MaterialState.disabled` when transforming the set into a list, so you would have to check all possible orders or sort the list. The easier workaround imho is to use plain old if statements.

**Comment by miDeb** [comment]
An interesting question I think is whether additional elements should be ignored, like with Map patterns, or if the elements have to match exactly, like with Lists (+ optional `...` element to match the rest). The example I gave kind of assumes that they work like Map patterns, so additional entries (e.g. `MaterialState.hovered`) shouldn't be taken into account.

**Comment by mateusfccp** [comment]
> I don't think this would work reliably, as there is no guarantee that `MaterialState.selected` will come before `MaterialState.disabled` when transforming the set into a list

AFAIK, both `toList()` and spread consider iteration order, so if `states` are ordered in a specific way, `states.toList()` or `[...states]` will be ordered the same way.

**Comment by miDeb** [comment]
@mateusfccp
- The `Set` interface does not specify that iteration order is stable (in practice a `LinkedHashSet` might be used, but that's not guaranteed)
- In this specific case it is unspecified in which order flutter adds those elements to the set, I can't make any assumptions about that.

My point wasn't that `.toList()` should be used instead of the spread operator, but that transforming a Set to a List and then using list patterns is not going to produce the intended behavior.

**Comment by lrhn** [comment]
A set pattern is *possible*, but it would necessarily differ from list patterns and map patterns.

A list pattern allows you to specify elements by (constant) index, or  (constant) distance from the end, a map allows you to specify elements by (constant) key.
They have an inherent structure that makes it possible to specify individual elements, and then match patterns against those elements.

Sets do not. They have no specified ordering (no indexing, no other keys), so it's not possible to apply a sub-pattern meaningfully to a single element.
A set is more like the *keys* of a map, than the values. More like a `Map<E, Null>` than anything else. 
(A `Map.keys` iterable is remarkably like an unmodifiable set, because a `Set` is an iterable collection with efficient `contains`, `length`, `add` and `remove`, so an unmodifable set is really just an iterable with efficient `contains` and `length`. And no specified order.)

So a set pattern could have the form `{constant, constant, constant}` or `{constant, constant, ...}`, with an optional type argument, `<type>{constant, constant, constant}`, one requiring the pattern to mention all the values, the other to allow only mentioning a subset of the values.
Or, choose the map pattern strategy of always having an implicit `...`, in which case it's just 
```ebnf
<setPattern> ::= <typeArgument>? '{' (<expression> (',' <expression>) ','?)? '}'
```
With `...` it's:
```dart
<setPattern> ::= <typeArgument>? '{' (<setPatternElements> ','?)? '}'
<setPatternElements> ::= 
      '...' 
    | <expression> (',' <expression>)* (',' '...')? '.'?
```
where the `<typeArgument>` must only have one type (otherwise it's possibly map pattern, or just a compile-time error),
and the `<expression>`s must be compile-time constant expressions, and are in a constant scope.

Since we don't know the *equality* used by the map, it's probably more honest to drop the `length` check entirely, and just do `.contains` on every mentioned element, and let the set sort them out.

The meaning of 
```dart
case {1, 2, 3}: /*....*/
case {1, 2, ...}: /*...*/
```
would be something like this desugaring (ignoring caching of member accesses for now):
```dart
if (_value is Set && _value.length == 3 && _value.contains(1) && value.contains(2) && value.contains(3)) {
  ...
} else if (_value is Set && _value.length >= 2 && _value.contains(1) && _value.contains(2)) {
  ...
}
```
We can drop the `_value.length` check in the second case. It's really just there as a quick bailout if it's not possible for the set to have all the required elements (but that also assumes that the set considers those elements distinct, which, e.g., `switch ({1}) { case {1, 1.0}: print("1 ==1.0"); }`  doesn't satisfy).

With implicit `...` and no length checks, it's just:
```dart
if (_value is Set && _value.contains(1) && value.contains(2) && value.contains(3)) {
  ...
} else if (_value is Set && _value.contains(1) && _value.contains(2)) {
  ...
}
```
(with lots of ways to optimize that.)

We'll probably have the same rules as for map keys: No two constants which are structurally equivalent is allowed.

**Comment by miDeb** [comment]
I think it would make sense for matching a Set to work like a `Map<T, void>` (because you could argue that a Set is just a Map with only keys) where a set pattern `{constant, constant, constant}` is the equivalent of a map pattern `{constant: _, constant: _, constant: _}`
